### PR TITLE
Add broadcast compatibility for utility functions

### PR DESCRIPTION
**Pull request by bruAristimunha**
## Summary

- Make all public utility functions in `base.py`, `mean.py`, `covariance.py`, and `tangentspace.py` handle arbitrary batch dimensions `(*batch, n, n)` via NumPy broadcasting, enabling batched computation without explicit loops at the call site.
- Replace Python loops over batch elements with vectorized NumPy operations for key functions (`cross_spectrum`, `coherence`, `covariance_mest`, `nearest_sym_pos_def`, etc.).
- Fix conjugate transpose calls, convergence norms, and edge cases (division by zero, empty batches) across all modified modules.

## Changes by module

- **`base.py`**: Vectorize `_first_divided_difference`, fix `ddexpm`/`ddlogm` for batched inputs, rewrite `nearest_sym_pos_def` with batch SVD
- **`mean.py`**: Add `_max_norm` helper, fix all mean functions (`riemann`, `logdet`, `wasserstein`, `thompson`, `power`, `chol`, `logchol`, `alm`, `ale`) for batch support
- **`covariance.py`**: Vectorize `covariance_scm`/`covariance_sch`, fix `_complex_estimator` axis, add flatten-loop-reshape wrappers for functions that need it
- **`tangentspace.py`**: Fix `_check_dimensions`, `exp_map_wasserstein`, and transport functions for batch dims
- **`tests/test_broadcast_compat.py`**: 146 tests covering 4D and 5D batch shapes across all modified functions

## Test plan

- [ ] All existing tests pass (no regressions)
- [ ] New broadcast compatibility tests pass (`pytest tests/test_broadcast_compat.py`)
- [ ] Verify 4D and 5D batch shapes work for all utility functions

### CODE REVIEW

File: pyriemann/utils/covariance.py [hunk 1]
```diff
@@ -196,22 +196,32 @@ def weight_func(x):  # Example 2, Section V-C in [1]
         raise ValueError(f"Unsupported m_estimator: {m_estimator}")
 
     if not assume_centered:
-        X -= np.mean(X, axis=1, keepdims=True)
+        X = X - np.mean(X, axis=-1, keepdims=True)
     if init is None:
-        cov = X @ X.conj().T / n_times
+        cov = X @ np.swapaxes(X.conj(), -2, -1) / n_times
```
Comment (qbarthelemy) [inline pyriemann/utils/covariance.py]:
```suggestion
        cov = X @ ctranspose(X) / n_times
```

File: pyriemann/utils/covariance.py
Comment (qbarthelemy) [inline pyriemann/utils/covariance.py]: ?
```suggestion
    C, _ = ledoit_wolf(np.swapaxes(X, -2, -1), **kwds)
```
Same question for ` _mcd` and `_oas`

File: pyriemann/utils/mean.py [hunk 2]
```diff
@@ -139,9 +170,8 @@ def mean_alm(X, *, tol=1e-14, maxiter=100, sample_weight=None, **kwargs):
             s = np.mod(np.arange(h, h + n_matrices - 1) + 1, n_matrices)
             M_iter[h] = mean_alm(M[s], sample_weight=sample_weight[s])
 
-        norm_iter = np.linalg.norm(M_iter[0] - M[0], ord=2)
-        norm_c = np.linalg.norm(M[0], ord=2)
-        if (norm_iter / norm_c) < tol:
+        crit = _batch_norm(M_iter[0] - M[0], ord=2) / _batch_norm(M[0], ord=2)
```
Comment (qbarthelemy) [inline pyriemann/utils/mean.py]: This will change the convergence of algorithms.
It would be necessary to prove that it's better.

File: pyriemann/utils/tangentspace.py [hunk 3]
```diff
@@ -223,12 +223,12 @@ def exp_map_wasserstein(X, Cref):
...
-    C = 1 / (d[:, None] + d[None, :])
+    C = 1 / (d[..., :, np.newaxis] + d[..., np.newaxis, :])
 
-    X_rotated = V.conj().T @ X @ V
+    X_rotated = ctranspose(V) @ X @ V
     X_tmp = C * X_rotated
-    X_tmp = X_tmp @ np.diag(d) @ X_tmp
-    X_tmp = V @ X_tmp @ V.conj().T
+    X_tmp = X_tmp @ (d[..., :, np.newaxis] * X_tmp)
+    X_tmp = V @ X_tmp @ ctranspose(V)
```
Comment (qbarthelemy) [inline pyriemann/utils/tangentspace.py]: `V` is nxn, so `ctranspose` is not necessary
```suggestion
    Vh = V.conj().T
    X_tmp = C * Vh @ X @ V
    X_tmp = X_tmp @ (d[..., :, np.newaxis] * X_tmp)
    X_tmp = V @ X_tmp @ Vh
```

File: pyriemann/utils/covariance.py [hunk 4]
```diff
@@ -58,21 +59,21 @@ def wrapper(X, **kwds):
 @_complex_estimator
 def _lwf(X, **kwds):
     """Wrapper for sklearn ledoit wolf covariance estimator"""
-    C, _ = ledoit_wolf(X.T, **kwds)
+    C, _ = ledoit_wolf(np.swapaxes(X, -2, -1), **kwds)
```
Comment (qbarthelemy) [inline pyriemann/utils/covariance.py]: I am sorry, but this modification is useless,
since input `X` of `ledoit_wolf` is `(n_samples, n_features)` instead of `(..., n_samples, n_features)`
https://scikit-learn.org/stable/modules/generated/sklearn.covariance.ledoit_wolf.html

Same remark for `_mcd` and `_oas`

File: pyriemann/utils/distance.py [hunk 5]
```diff
@@ -991,8 +991,8 @@ def distance_mahalanobis(X, cov, mean=None, squared=False):
     .. [1] https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.mahalanobis.html
     """  # noqa
     if mean is not None:
-        X -= mean
+        X = X - mean
 
     Xw = invsqrtm(cov) @ X
-    d2 = np.einsum("ij,ji->i", Xw.conj().T, Xw).real
+    d2 = np.sum(np.abs(Xw)**2, axis=-2).real
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply pyriemann/utils/distance.py]: Kept .real for dtype safety — while np.abs(Xw)**2 is mathematically real, complex-valued inputs may retain complex dtype in the sum. This matches the pattern in the original einsum version.

File: pyriemann/utils/mean.py [hunk 6]
```diff
@@ -321,27 +350,28 @@ def mean_logchol(X, sample_weight=None, **kwargs):
...
-    n_matrices, _, n_channels = X.shape
+    n_matrices = X.shape[0]
+    n_channels = X.shape[-1]
     sample_weight = check_weights(sample_weight, n_matrices)
 
     X_chol = np.linalg.cholesky(X)
-    L = np.zeros(X.shape[-2:], dtype=X.dtype)
+    L = np.zeros(X.shape[1:], dtype=X.dtype)
 
     tri0, tri1 = np.tril_indices(n_channels, -1)
```
Comment (qbarthelemy) [inline pyriemann/utils/mean.py]:
```suggestion
    tri0, tri1 = np.tril_indices(n, -1)
```

File: tests/conftest.py [hunk 7]
```diff
@@ -62,3 +62,32 @@ def _gen_targets(n_matrices):
         return np.random.rand(n_matrices)
 
     return _gen_targets
+
+
+# ---- Broadcast compatibility test helpers ----
+
+BATCH_SHAPES = [(2, 3), (2, 3, 4)]
```
Comment (qbarthelemy) [inline tests/conftest.py]: I think that the size of the batches should not be identical for each test, but should vary for each test (like n_matrices and n_channels), in order to have larger and therefore safer tests.

File: pyriemann/utils/mean.py [hunk 8]
```diff
@@ -53,7 +70,21 @@ def mean_ale(X, *, tol=10e-7, maxiter=50, sample_weight=None, init=None):
         <https://arxiv.org/abs/1505.07343>`_
         M. Congedo, B. Afsari, A. Barachant, M. Moakher. PLOS ONE, 2015
     """
-    n_matrices, n, _ = X.shape
+    n_matrices = X.shape[0]
```
[the remark this answers is not in the document]
Comment (qbarthelemy) [reply pyriemann/utils/mean.py]: Can you give me an example of functions in scikit-learn that handle inputs of shape (n1, ..., n2)?

In NumPy, SciPy and sklearn, when there are additional dimensions, they are always the leftmost ones, ie shape (..., n1, n2),
For example, see https://scikit-learn.org/stable/modules/generated/sklearn.covariance.shrunk_covariance.html

File: tests/test_utils_distance.py [hunk 9]
```diff
@@ -397,3 +400,35 @@ def test_distance_mahalanobis_scipy(mean, get_mats):
...
+    distance_riemann,
+    distance_wasserstein,
+    distance_kullback,
+    distance_kullback_right,
+    distance_kullback_sym,
+    distance_logdet,
+    distance_thompson,
+    pytest.param(partial(distance_poweuclid, p=0.5), id="distance_poweuclid"),
+])
+def test_distance_pair_broadcast(func, batch_shape):
```
Comment (qbarthelemy) [inline tests/test_utils_distance.py]: Why not complete `test_distance_ndarray`?

**Review by qbarthelemy — CHANGES_REQUESTED** [review]

**Review by agramfort — APPROVED** [review]
Assuming there is no bug fix in there which would require a dedicated PR
